design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,613 @@
+authenticator executions under the authentication factor are considered valid alternatives,
+and are evaluated in order. The authentication stops being automatically evaluated at the
+first authentication execution that requires user input (alternative: all non-user input
+authentication executions are evaluated before the first user-input authentication execution).
+
+Adding a 2nd "authentication factor" execution sets the 2nd factor, and it must then have
+authentication executions assigned. To have an authenticator execution be valid for both
+authentication factors it must be set under both 1st and 2nd authentication factor. For
+example, kerberos could be set for both 1st and 2nd factor, which would mean that the user
+skips both factors if he's registered with kerberos, but it could also be just set for the
+first factor, in which case he'd still have to input the 2nd factor. To handle optional 2nd
+factors there could either be a "optional authentication factor type" or have an
+"optional" flag in the options of the "authentication factor".
+
+Potentially, this system could also allow a company that's really security
+conscientious/paranoiac to set N factors.
+By setting a single authentication factor with an authentication executions that are not
+passwords (e.g. Fido2), this system also allows passwordless login.
+
+Authentication factors are treated as a bloc in the evaluation of alternatives. That is, if in
+a flow there's "Identity provider redirector", "authentication factor 1", "authentication
+factor 2", entering the authentication factor 1 flow will automatically cause "authentication
+factor 2" to be executed after.
+
+To make things perhaps a little more clear, the current default "Browser" authentication
+would become for example:
+
+```
+Auth type
+------------------------------------------------------------
+Identity Provider Redirector
+Authentication factor (1)
+  |- Cookie
+  |- Kerberos
+  |- Username Password Form
+Authentication factor (2) [OPTIONAL]
+  |- Cookie
+  |- Kerberos
+  |- OTP Form
+```
+And if we wanted to have an mandatory 2nd factor, with either OTP or a FIDO2 configured:
+
+```
+Auth type
+------------------------------------------------------------
+Identity Provider Redirector
+Authentication factor (1)
+  |- Cookie
+  |- Kerberos
+  |- Username Password Form
+Authentication factor (2)
+  |- Cookie
+  |- Kerberos
+  |- OTP Form
+  |- FIDO-2
+```
+Potentially we could also introduce another type of authentication execution, we could call it "Authenticated on", to simplify the authenticators that bypass the authentication factors. We could then have:
+
+```
+Auth type
+------------------------------------------------------------
+Identity Provider Redirector
+Authenticated on
+  |- Cookie
+  |- Kerberos
+Authentication factor (1)
+  |- Username Password Form
+Authentication factor (2)
+  |- OTP Form
+  |- FIDO-2
+```
+
+#### Authentication categories
```

Comment:
I somewhat follow the idea of authentication categories and the need for something like it. There's some work to be done here though to expand on this concept to complete it. How are they created? How are they used? Etc.

Reply:
I'll describe this in greater detail. My main dilemma was, "are authentication categories something statically created in the code?", or "are authentication categories something that can be created and managed by an administrator, with a database presence?". Comments up until now indicate the latter.

Reply:
Perhaps in reality its both? One thing I'm not to keen on with the current flows is that the "static" data from code is copied into the DB. It makes upgrades complex.